src/libsyntax/ext/expand.rs

```diff
@@ -673,22 +738,28 @@ impl<'feat> ExpansionConfig<'feat> {
     }
 }
 
-pub fn expand_crate(mut cx: ExtCtxt,
-                    user_exts: Vec<NamedSyntaxExtension>,
-                    mut c: Crate) -> (Crate, HashSet<Name>) {
-    if std_inject::no_core(&c) {
+// Sets crate root and inserts user extensions into the context.
+fn init_cx(cx: &mut ExtCtxt, user_exts: Vec<NamedSyntaxExtension>, c: &Crate) {
+    if std_inject::no_core(c) {
         cx.crate_root = None;
-    } else if std_inject::no_std(&c) {
+    } else if std_inject::no_std(c) {
         cx.crate_root = Some("core");
     } else {
         cx.crate_root = Some("std");
     }
-    let ret = {
-        let mut expander = MacroExpander::new(&mut cx);
 
-        for (name, extension) in user_exts {
-            expander.cx.syntax_env.insert(name, extension);
-        }
+    for (name, extension) in user_exts {
+        cx.syntax_env.insert(name, extension);
+    }
```

Comment:
I think we should load these user-defined extensions in the same place we load ordinary macros (currently, `expander.load_macros(&items);` in `expand_crate`).